### PR TITLE
Print multiline strings

### PR DESCRIPTION
**Pull request by christopher-dG**
As discussed in #91, there are some cases where multiline strings are beneficial.
With this implementation, strings are printed on multiple lines if:

- they have newlines which are not leading or trailing
- they contain double quotes

I'm using YAML primarily for BrokenRecord.jl where we store HTTP response bodies, which are often JSON. So both cases listed above are very common, and they look much much better as a multiline string that doesn't require escaping.

This is very much open for discussion, if people prefer the simplicity of `repr` then we can, then we can stick with that. But I'd argue that  most people use YAML because it's meant to be human-readable :)

### CODE REVIEW

**Comment by christopher-dG** [comment]
The one thing I do not like about using multiline strings is what happens when they are the last key in a file.

For example:

```yml
a: true
b: |
  hello









```

This file has a bunch of trailing newlines, many people (like me) have their editors set up to remove trailing whitespace from their files. I think you can mark the end of a YAML document with a `---` line, perhaps it's a good idea to always put that at the end?

**Review by mirkobunse — COMMENTED** [review]
Thanks for the contribution! Looks good to me.

Only thing I'm curious to hear about is why you don't use `repr` when `"` occurs in `str` (see my comment in line 82 of `src/writer.jl`).

I guess Kevin needs to decide whether multi-line handling should become the one-and-only type of writing yaml, or whether we need a keyword argument `multiline` to make the user choose between a pure `repr`-based printing (like before) and your multiline-solution.

**Comment by mirkobunse** [comment]
> The one thing I do not like about using multiline strings is what happens when they are the last key in a file.

Actually, I do like what happens in this case, as long as the roundtrip julia/yaml/julia is producing identical results :)

File: src/writer.jl [hunk 1]
```diff
@@ -79,7 +79,22 @@ end
 
 # _print a single string
 _print(io::IO, str::AbstractString, level::Int=0, ignore_level::Bool=false) =
-    println(io, repr(MIME("text/plain"), str)) # quote and escape
+    if occursin('\n', strip(str)) || occursin('"', str)
```
[the remark this answers is not in the document]
Comment (christopher-dG) [reply src/writer.jl:82]: Consider the following string:

```
{"a":"b","c":"d","e":"f"}
```

`repr` on this is `"{\"a\":\"b\",\"c\":\"d\",\"e\":\"f\"}"`, which is much less readable than multiline version.

**Comment by christopher-dG** [comment]
The trailing newlines come from `b`. The only problem I see with it is it's easy to accidentally strip them if your editor is configured to do so.

File: test/runtests.jl [hunk 2]
```diff
@@ -224,4 +224,11 @@ order_two = OrderedDict(dict_content[[2,1]]...) # reverse order
 @test YAML.load(YAML.yaml(Dict("a" => """a "quoted" string""")))["a"] == """a "quoted" string"""
 @test YAML.load(YAML.yaml(Dict("a" => """a \\"quoted\\" string""")))["a"] == """a \\"quoted\\" string"""
 
+@test YAML.load(YAML.yaml(Dict("a" => "")))["a"] == ""
+@test YAML.load(YAML.yaml(Dict("a" => "nl at end\n")))["a"] == "nl at end\n"
+@test YAML.load(YAML.yaml(Dict("a" => "one\nnl\n")))["a"] == "one\nnl\n"
+@test YAML.load(YAML.yaml(Dict("a" => "many\nnls\n\n\n")))["a"] == "many\nnls\n\n\n"
+@test YAML.load(YAML.yaml(Dict("a" => "no\ntrailing\nnls")))["a"] == "no\ntrailing\nnls"
+@test YAML.load(YAML.yaml(Dict("a" => "foo\n\"bar\\'")))["a"] == "foo\n\"bar\\'"
+
```
[the remark this answers is not in the document]
Comment (christopher-dG) [reply test/runtests.jl:233]: Yeah, the format is very finnicky 😅

**Comment by DilumAluthge** [comment]
> The trailing newlines come from `b`. The only problem I see with it is it's easy to accidentally strip them if your editor is configured to do so.

Yeah e.g. my editor (Atom) will remove the trailing newlines.

Any reason why we shouldn't just put the `---` at the end?

**Comment by christopher-dG** [comment]
I support always adding the `---`. I think it'll in those edge cases both to prevent accidentally editing of the file, and to make it more clear to readers that such a string has trailing newlines.

File: Project.toml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 name = "YAML"
 uuid = "ddb6d928-2868-570f-bddf-ab3f9cf99eb6"
-version = "0.4.3"
+version = "0.4.4"
```
Comment (kescobo) [inline Project.toml:3]: Does it make sense to have version bumps occur in PRs? In my own repos, I tend to make separate commits where the only thing that happens is a version bump.

Doesn't make a huge difference to me, except that, in a case where multiple PRs are happening at the same time, there may be conflicts.

**Comment by kescobo** [comment]
> I guess Kevin needs to decide whether multi-line handling should become the one-and-only type of writing yaml or whether we need a keyword argument multiline to make the user choose between a pure repr-based printing (like before) and your multiline-solution.

Oh no! Why is this on me?!? :rofl: 

I think having a `multiline` keyword would be beneficial, the question is what the best default would be - `true` seems the most sensible. I also wonder if having the trailing `---` could be an option as well. Ideally, we'd have a situation where any (valid) input could be round-tripped and produce identical output, though of course this isn't always feasible given the flexibility of the format. 

Making these options doesn't have to be part of this PR in my opinion, I'm fine with multi-line and trailing `---` being the defaults.

Though I should say in all seriousness, it probably doesn't make sense for me to be the one making these decisions, since I rarely use this package - my stewardship is a somewhat strange accident. In the absence of any objection, I'm happy to let y'all have final say.

File: Project.toml (hunk 3 above)
Comment (christopher-dG) [reply Project.toml:3]: I usually do it. It doesn't conflict if other PRs bump to the same version. But I can totally remove it if you want!

**Comment by kescobo** [comment]
The error in CI is

```
test = spec-02-13: Error During Test at /home/travis/build/JuliaData/YAML.jl/test/runtests.jl:139

  Got exception outside of a @test

  ArgumentError: `nothing` should not be printed; use `show`, `repr`, or custom output instead.

  Stacktrace:

   [1] print(::IOContext{Base.GenericIOBuffer{Array{UInt8,1}}}, ::Nothing) at ./show.jl:566

   [2] print(::IOContext{Base.GenericIOBuffer{Array{UInt8,1}}}, ::String, ::String, ::Vararg{Any,N} where N) at ./strings/io.jl:42

   [3] show(::IOContext{Base.GenericIOBuffer{Array{UInt8,1}}}, ::YAML.ParserError) at /home/travis/build/JuliaData/YAML.jl/src/parser.jl:25
# ...
```

**Comment by christopher-dG** [comment]
Looks like a valid failure, I didn't even realize that you could have strings at the top level...

**Comment by christopher-dG** [comment]
Also, I read the spec and we should actually be using `...` to end a document if we decide to go that route.

**Comment by kescobo** [comment]
Good to know! But this PR is ready to merge? Maybe worth opening issues for optional single-line printing and including `...` at the end so we can keep track of it?

**Comment by christopher-dG** [comment]
Yeah, this is good to go.
